QuizApp/Search/SearchQuizViewController.swift

```diff
@@ -0,0 +1,112 @@
+        self.init()
+        self.router = router
+        self.quizzesUseCase = quizzesUseCase
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        buildViews()
+        addConstraints()
+    }
+
+    func buildViews() {
+        view.backgroundColor = .systemBlue
+        
+        tableView = UITableView()
+        tableView.rowHeight = 150
+        tableView.register(CustomTableViewCell.self, forCellReuseIdentifier: cellIdentifier)
+        tableView.register(TableHeader.self, forHeaderFooterViewReuseIdentifier: headerIdentifier)
+        tableView.isUserInteractionEnabled = true
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.backgroundColor = .clear
+        
+        searchBarView = SearchBarView()
+        searchBarView.searchButton.addTarget(self, action: #selector(searchQuizzes), for: .touchUpInside)
+        
+        view.addSubview(tableView)
+        
+        tableView.tableHeaderView = searchBarView
+    }
+
+    private func addConstraints() {
+        tableView.translatesAutoresizingMaskIntoConstraints = false
+        tableView.autoPinEdge(toSuperviewSafeArea: .leading, withInset: 10)
+        tableView.autoPinEdge(toSuperviewSafeArea: .trailing, withInset: 10)
+        tableView.autoPinEdge(toSuperviewSafeArea: .top, withInset: 30)
+        tableView.autoPinEdge(toSuperviewEdge: .bottom)
+    }
+    
+    private func groupByCategory(quizzesList: [Quiz]) -> [(QuizCategory,Array<Quiz>)] {
+        var quizzesGroupedByCategory = Dictionary<QuizCategory, Array<Quiz>>()
+        for quiz in quizzesList {
+            if quizzesGroupedByCategory[quiz.category] == nil {
+            quizzesGroupedByCategory[quiz.category] = Array<Quiz>()
+          }
+            quizzesGroupedByCategory[quiz.category]?.append(quiz)
+        }
+        return quizzesGroupedByCategory.compactMap{$0}
+      }
+    
+    @objc
+    func searchQuizzes() {
+        let quizzes = quizzesUseCase.filterQuizzes(name: searchBarView.searchBar.text)
+        quizzesGroupedByCategory = groupByCategory(quizzesList: quizzes)
+        tableView.reloadData()
+        }
+}
+
+extension SearchQuizViewController: UITableViewDataSource, UITableViewDelegate {
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return quizzesGroupedByCategory.count
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return quizzesGroupedByCategory[section].1.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as! CustomTableViewCell
+        let quiz = quizzesGroupedByCategory[indexPath.section].1[indexPath.row]
+        let url = URL(string: quiz.imageUrl)
+        let data = try? Data(contentsOf: url!)
+        cell.imgQuiz.image = UIImage(data: data!)
```

Comment:
This will cause a crash if there is no internet. Also,`Data(contentsOf: url!)` is synchronous which means it will block the thread until it gets the image. You might not notice it, but in slow internet conditions and with more cells you definitely would.
This could be fixed by making the cell handle the image download asynchronously and without force unwrapping.